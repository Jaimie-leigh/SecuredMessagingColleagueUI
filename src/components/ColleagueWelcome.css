.colleagueWelcome {
  align-content: center;
}

.colleagueDetails {
  display: flex;
  background-color: #2d8259;
  flex: 1;
  color: white;
}

.colleagueID {
  flex: 1;
  padding-left: 0.5em;
}

.signOut a {
  color: white;
}

.colleagueDetails button:hover {
  background-color: white;
  color: black;
}

.helpText {
  font-weight: inherit;
}

.appName {
  flex: 1.2;
}

.colleagueDetails button {
  min-width: fit-content;
  min-height: 100%;
  margin: 0;
  background-color: #2178b0;
  font-weight: bold;
}

.infoBox {
  width: max-content;
  padding: 1em;
  text-align: center;
  width: -webkit-fill-available;
}

.brokerMessages {
  padding: 2em;
}

.brokerMessageHeaders {
  background-color: #edf6ff;
  color: #071d49;
  display: flex;
  align-items: center;
  font-weight: bold;
  padding: 0.5em;
  border-top: solid 2px #005eb8;
  border-bottom: solid 1px #005eb8;
}

.innerMessage {
  flex: 1;
}

.innerMessage-2 {
  flex: 2;
}

.innerMessage:last-child {
  text-align: right;
}

.innerMessage button {
  min-width: auto;
  background-color: #77ba00;
}

.brokerMessageDetails {
  background-color: white;
  display: flex;
  align-items: center;
  font-weight: bold;
  padding: 0.5em;
  border-bottom: solid 1px #b5bbc8;
}

.sendMessageForm {
  background-color: white;
  padding: 1em;
}

.messageSent {
  background-color: #2178b0;
    color: white;
    padding: 0.1em;
    margin-top: 0.5em;
    text-align: center;
}