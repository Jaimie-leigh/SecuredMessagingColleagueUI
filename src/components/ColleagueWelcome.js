import React from "react";
import "./ColleagueWelcome.css";
import "./loading.css";
import FadeLoader from "react-spinners/FadeLoader";
import Moment from "moment";
import { ReactComponent as DownIcon } from "../images/down-chevron.svg";
import { ReactComponent as UpIcon } from "../images/up-chevron.svg";
import { Form, Button } from "react-bootstrap";

class ColleagueWelcome extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      data: [],
      messageData: [],
      newMessages: [],
      isLoading: false,
      selectedSubjectId: "",
      selectedChainId: "",
      formMessageBody: "",
      messageSent: false,
    };

    this.handleClick = this.handleClick.bind(this);
    this.handleFormSubmit = this.handleFormSubmit.bind(this);
  }

  componentDidMount() {
    this.setState({ isLoading: true });
    const serachByColleagueId = this.props.location.state;
    const jaisProxyUrl = "https://nameless-hamlet-06819.herokuapp.com/";
    const url =
      "http://securedmessaging.azurewebsites.net/api/Colleague/" +
      serachByColleagueId;

    fetch(jaisProxyUrl + url)
      .then((res) => res.json())
      .then((data) => this.setState({ data: data }))
      .then(
        fetch(
          jaisProxyUrl +
            "http://securedmessaging.azurewebsites.net/api/ApplicationMessages"
        )
          .then((res) => res.json())
          .then((messageData) =>
            this.setState(
              { messageData: messageData, isLoading: false },
              () => {}
            )
          )
      );
  }

  routeChange() {
    window.location.href = "/";
  }

  //saves the message body to state on change 
  handleBodyChange = (event) => {
    this.setState({ formMessageBody: event.target.value });
  };

  // sets the subject and chain id for the messgae the colleague is replying too 
  handleClick(message) {

    if(this.state.selectedSubjectId !== message.message_Subject.subjectId)
    {
    this.setState(
      { selectedSubjectId: message.message_Subject.subjectId } , () => {
        });
      } else {
        this.setState(
          { selectedSubjectId: null } , () => {
            });
      }

      if(this.state.selectedChainId !== message.message_Subject.message_Chain.messageChainId) {
    this.setState(
          { selectedChainId: message.message_Subject.message_Chain.messageChainId },
          () => {
            //callback to set state instantly
          }
        );
      } else {
        this.setState(
          { selectedChainId: null },
          () => {
            //callback to set state instantly
          })
      };
  }

  // handles user submiting the message from and sends post request to API
  // shows loading screen when message submit. sends message to API and reloads page
  handleFormSubmit(e) {
    this.setState({ isLoading: true });
    e.preventDefault();

    // set next chainID from string +1 with need int
    let chainID = parseInt(this.state.selectedChainId) +1;
    chainID = chainID.toString();
    // get current time to add to post body 'date and time'
    const submit_time = Moment().format("ddd DD MMM YYYY HH:mm:ss");

    const jaisProxyUrl = "https://nameless-hamlet-06819.herokuapp.com/";
    try {
      fetch(
        jaisProxyUrl +
          "http://securedmessaging.azurewebsites.net/api/Message_Chain",
        {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          // convert the React state to JSON and send it as the POST body
          body: JSON.stringify({
            messageChainId: chainID,
            messageSubjectId: this.state.selectedSubjectId,
            messageBody: this.state.formMessageBody,
            sentFromId: this.props.location.state,
            dateTime: submit_time,
          }),
        }
      )
        .then((data) =>
          this.setState(
            { messageSent: true, isLoading: false},
            this.componentDidMount()
          )
        );
     } catch (error) {};
  };

  // renders to display on page
  render() {
    const { data, isLoading, messageData } = this.state;
    // if page is loading displays loading text and spinner to make user awear 
    if (isLoading) {
      return (
        <div className="pageLoading">
          <p>Loading...</p>
          <FadeLoader size={150} color={"#2d8259"} loading={isLoading} />
        </div>
      );
    }

    // If colleague has not been found returns user not found text
    if (!data.colleagueForename) {
      return <p>User not found</p>;
      // needs a retun button and a bit more info
    }

    // else return as valid colleague and display, if any, message data
    return (
      <div className="colleagueWelcome">
        <div className="colleagueDetails">
          <p className="colleagueID">
            <strong>
              Welcome {data.colleagueForename} {data.colleagueSurname}
            </strong>
          </p>
          <p className="appName">
            <strong>Secured Messaging</strong>
          </p>
          <button className="signOut" onClick={() => this.routeChange()}>
            Sign Out
          </button>
        </div>
        {this.state.messageSent === true && (
          <div className="messageSent">{this.displaySentText()}</div>
        )}
        <div className="infoBox">
          <p>
            <strong>Message Backlog</strong>
          </p>
          <p className="helpText">
            The below application have open chats awaiting a response. Click on
            the roll number to send and view messages relation to the
            application.
          </p>
        </div>
        <div className="brokerMessages">
          <div className="brokerMessageHeaders">
            <div className="innerMessage">Roll Number</div>
            <div className="innerMessage">Subject</div>
            <div className="innerMessage">Data and Time</div>
            <div className="innerMessage-2">Message</div>
            <div className="innerMessage">Reply to Message</div>
          </div>
          <div>{this.getMessages(messageData)}</div>
        </div>
      </div>
    );
  }


  // gets all application messages and checks which messages are from a broker which have not yet been replied too,
  // saves the new messages into an array 
  getMessages(messageData) {
    var newMessagesArray = [];
    var latestMessageIndex;
    var latestMessage;

    if(messageData.message_Subjects){
    messageData.message_Subjects.forEach(saveNewBrokerMessage)
    }

    // saves the new messages into an array with the rquired details 
    function saveNewBrokerMessage(sub) {
      latestMessageIndex = sub.message_Chain.length - 1;
      latestMessage = sub.message_Chain[latestMessageIndex];
      if (latestMessage.sentFromId.toString().length === 7) {
        var message_Chain = {
          dateTime: latestMessage.dateTime,
          sentFromId: latestMessage.sentFromId,
          messageBody: latestMessage.messageBody,
          messageChainId: latestMessage.messageChainId,
        };
        var message_subject = {
          brokerId: sub.brokerId,
          subjectId: sub.messageSubjectId,
          subject: sub.subject,
          message_Chain: message_Chain,
          rollNumber: sub.rollNumber,
        };
        newMessagesArray.push({ message_Subject: message_subject });
      } else {
        // do nothing as latest message was not from a broker
      }
    }

    // if there are messages in the array, display the broker message
    if (newMessagesArray.length > 0) {
      return (newMessagesArray = newMessagesArray.map((item) => (
        <div>
          <div className="brokerMessageDetails"  key={item}>
            <div className="innerMessage" key={item.message_Subject.rollNumber}>
              {item.message_Subject.rollNumber}
            </div>
            <div className="innerMessage" key={item.message_Subject.subject}>
              {item.message_Subject.subject}
            </div>
            <div className="innerMessage" key={item.message_Subject.message_Chain.dateTime}>
              {Moment(item.message_Subject.message_Chain.dateTime).format(
                "DD MMM YY HH:mm"
              )}
            </div>
            <div className="innerMessage-2" key={item.message_Subject.message_Chain.messageBody}>
              {item.message_Subject.message_Chain.messageBody}
            </div>
            <div className="innerMessage">
              <button onClick={() => this.handleClick(item)}>
                {this.state.selectedSubjectId ===
                item.message_Subject.subjectId ? (
                  <UpIcon className="showHideIcons" />
                ) : (
                  <DownIcon className="showHideIcons" />
                )}
              </button>
            </div>
          </div>
          <div>
            {this.state.selectedSubjectId ===
              item.message_Subject.subjectId && (
              <Form
                className="sendMessageForm"
                onSubmit={this.handleFormSubmit}
              >
                <Form.Group>
                  <Form.Label>
                    Enter message to reply
                    <Form.Control
                      as="textarea"
                      rows={3}
                      value={this.state.value}
                      onChange={this.handleBodyChange}
                      required
                      maxLength="250"
                      minLength="10"
                    />
                  </Form.Label>
                </Form.Group>
                <Button variant="primary" type="submit">
                  Send Message
                </Button>
              </Form>
            )}
          </div>
        </div>
      )));
    } else {
      return <p>No New messages.</p>;
    }
  }

  // once the usr has responded to a message a text bar appears to let them know their message has been sent
  displaySentText() {
    return (
      <div>
        <p>Your message has been sent!</p>
      </div>
    );
  }

}

export default ColleagueWelcome;
