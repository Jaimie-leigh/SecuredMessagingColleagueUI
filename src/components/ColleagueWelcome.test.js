import React, { useState } from "react";
import renderer from "react-test-renderer";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import ColleagueWelcome from "./ColleagueWelcome";

let container = null;

beforeEach(() => {
  // set a DOM elemant as render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // clean up on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe("Colleague Welcome tests", () => {
  test("test renders as per snapshot", () => {
    const colleagueId = "912345";
    const tree = renderer
      .create(<ColleagueWelcome location={colleagueId}></ColleagueWelcome>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("test page renders as epxected when loading", () => {
    const colleagueId = "912345";
    render(
      <ColleagueWelcome location={colleagueId}></ColleagueWelcome>,
      container
    );
    expect(container.innerHTML).toContain("Loading...");
  });

  test("mock fetch call, user not valid", async () => {
    const fakeResponse = [];
    const colleagueId = "91234555";

    jest.spyOn(window, "fetch").mockImplementation(() => {
      const fetchResponse = {
        json: () => Promise.resolve(fakeResponse),
      };
      return Promise.resolve(fetchResponse);
    });

    await act(async () => {
      render(<ColleagueWelcome location={colleagueId} />, container);
    });

    expect(container.innerHTML).toContain("User not found");

    window.fetch.mockRestore();
  });

  test("mock fetch call, valid responce from db should return applications", async () => {
    const fakeResponse = 
      {
        colleagueId: 912345,
        colleagueForename: "Rebecca",
        colleagueSurname: "Jones",
      }
    ;
    const colleagueId = "912345";

    jest.spyOn(window, "fetch").mockImplementation(() => {
      const fetchResponse = {
        json: () => Promise.resolve(fakeResponse),
      };
      return Promise.resolve(fetchResponse);
    });

    await act(async () => {
      render(<ColleagueWelcome location={colleagueId} />, container);
    });

    expect(container.innerHTML).toContain(
      "Welcome Rebecca Jones"
    );

    window.fetch.mockRestore();
  });
});
