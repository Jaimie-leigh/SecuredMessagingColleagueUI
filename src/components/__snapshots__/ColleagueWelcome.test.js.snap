// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Colleague Welcome tests test renders as per snapshot 1`] = `
<div
  className="pageLoading"
>
  <p>
    Loading...
  </p>
  <div
    className="css-1ujetdd"
  >
    <div
      className="css-1938t4y"
    />
    <div
      className="css-iguij7"
    />
    <div
      className="css-1y6778f"
    />
    <div
      className="css-dtp5nv"
    />
    <div
      className="css-1irslik"
    />
    <div
      className="css-1ro0oci"
    />
    <div
      className="css-1f7jqjv"
    />
    <div
      className="css-q5ow2x"
    />
  </div>
</div>
`;
